src/python/pants/backend/python/rules/python_test_runner.py

```diff
@@ -96,37 +102,86 @@ def get_packages_to_cover(
+
+    # TODO(John Sirois): PexInterpreterConstraints are gathered in the same way by the
+    #  `create_pex_from_target_closure` rule, factor up.
+    transitive_hydrated_targets = await Get[TransitiveHydratedTargets](Addresses, test_addresses)
     all_targets = transitive_hydrated_targets.closure
+    all_target_adaptors = [t.adaptor for t in all_targets]
+    interpreter_constraints = PexInterpreterConstraints.create_from_adaptors(
+        adaptors=all_target_adaptors, python_setup=python_setup
+    )
 
-    resolved_requirements_pex = await Get[Pex](
+    # Ensure all pexes we merge via PEX_PATH to form the test runner use the interpreter constraints
+    # of the tests. This is handled by CreatePexFromTargetClosure, but we must pass this through for
+    # CreatePex requests.
+    create_pex = functools.partial(CreatePex, interpreter_constraints=interpreter_constraints)
+
+    # NB: We set `--not-zip-safe` because Pytest plugin discovery, which uses
+    # `importlib_metadata` and thus `zipp`, does not play nicely when doing import magic
+    # directly
+    # from zip files. `zipp` has pathologically bad behavior with large zipfiles.
+    # TODO: this does have a performance cost as the pex must now be expanded to disk. Long
+    #  term,
+    # it would be better to fix Zipp (whose fix would then need to be used by importlib_metadata
+    # and then by Pytest). See https://github.com/jaraco/zipp/pull/26.
+    additional_args_for_pytest = ("--not-zip-safe",)
+
+    pytest_pex = await Get[Pex](
+        CreatePex,
+        create_pex(
+            output_filename="pytest.pex",
+            requirements=PexRequirements(requirements=pytest.get_requirement_strings()),
+            additional_args=additional_args_for_pytest,
+        ),
+    )
+
+    requirements_pex = await Get[Pex](
         CreatePexFromTargetClosure(
-            addresses=Addresses((adaptor.address,)),
-            output_filename="pytest-with-requirements.pex",
-            entry_point="pytest:main",
-            additional_requirements=pytest.get_requirement_strings(),
-            # NB: We set `--not-zip-safe` because Pytest plugin discovery, which uses
-            # `importlib_metadata` and thus `zipp`, does not play nicely when doing import magic directly
-            # from zip files. `zipp` has pathologically bad behavior with large zipfiles.
-            # TODO: this does have a performance cost as the pex must now be expanded to disk. Long term,
-            # it would be better to fix Zipp (whose fix would then need to be used by importlib_metadata
-            # and then by Pytest). See https://github.com/jaraco/zipp/pull/26.
-            additional_args=("--not-zip-safe",),
+            addresses=test_addresses,
+            output_filename="requirements.pex",
             include_source_files=False,
+            additional_args=additional_args_for_pytest,
         )
     )
 
+    test_runner_pex = await Get[Pex](
```

Comment:
Consider adding a comment explaining why we do not put sources into this Pex, unlike how we handle V1.

Reply:
V1 does not put sources in this pex, it puts sources in a sources pex, requirements in a requirements pex and the pytest pex merges both of those in:
https://github.com/pantsbuild/pants/blob/29cf9fc8b25a540f32e4a9e0aeeaaa6ea4dcc48f/src/python/pants/backend/python/tasks/python_execution_task_base.py#L163-L166
https://github.com/pantsbuild/pants/blob/29cf9fc8b25a540f32e4a9e0aeeaaa6ea4dcc48f/src/python/pants/backend/python/tasks/python_execution_task_base.py#L179-L184

In V2 we now built a pytest pex and a requirements pex and merge both of those using a PEX_PATH only entrypoint pex. I did it this way just to make breaking out the ordering of the PEX_PATH easy to see and change, ie: pytest then requirements.

We could put the sources in a Pex in v2 and there should be little perf difference. Without that, if sources change we push and pull a new source digest. With that we would create a new source pex and then pull that. I doubt the speed difference is noticeable. That said, its conceptually clear not to create a source pex since basre files are shipped with digests in V2, so that's the good reason in my mind to stick with how sources are currently handled in V2.

Reply:
To clarify, absolutely I agree that the current approach makes conceptual sense. There is no need to create a sources PEX because we have an alternative "native"/idiomatic way to populate a chroot with source files.

My comment was solely a suggestion to briefly explain the divergence from V1.

Although, now that I write that out, I take that back..our point is to move away from V1 because we now have a superior way of doing things. V2 should be the de facto, and, if anything, the onus would be on V1 to explain how it diverges from V2.